Remote configuration for the SDK proposal (#121)
open-telemetry/oteps: text/0121-config-service.md

> # A Dynamic Configuration Service for the SDK

Reviewer: Consider to move this to `experimental/...` this way we know that it is experimental.

Author: Done! @SergeyKanzhelev, would it be a good idea to move the z-pages otep there as well?

# A Dynamic Configuration Service for the SDK

This proposal is for an experiment to configure metric collection periods. Per-metric and tracing configuration is also intended to be added, with details left for a later iteration.

It is related to [this pull request](https://github.com/open-telemetry/opentelemetry-proto/pull/155)

## Motivation

During normal use, users may wish to collect metrics every 10 minutes. Later, while investigating a production issue, the same user could easily increase information available for debugging by reconfiguring some of their processes to collect metrics every 30 seconds. Because this change is centralized and does not require redeploying with new configurations, there is lower friction and risk in updating the configurations.

## Explanation

> This OTEP is a request for an experimental feature [open-telemetry/opentelemetry-specification#62](https://github.com/open-telemetry/opentelemetry-specification/pull/632). It is intended to seek approval to develop a proof of concept. This means no development will be done inside either the Open Telemetry SDK or the collector. Since this will be implemented in [opentelemetry-go-contrib](https://github.com/open-telemetry/opentelemetry-go-contrib) and [opentelemetry-collector-contrib](https://github.com/open-telemetry/opentelemetry-collector-contrib), all of this functionality will be optional.

Reviewer 2: ```suggestion
This OTEP is a request for an experimental feature [open-telemetry/opentelemetry-specification#62](https://github.com/open-telemetry/opentelemetry-specification/pull/632). It is intended to seek approval to develop a proof of concept. This means no development will be done inside either the OpenTelemetry SDK or the collector. Since this will be implemented in [opentelemetry-go-contrib](https://github.com/open-telemetry/opentelemetry-go-contrib) and [opentelemetry-collector-contrib](https://github.com/open-telemetry/opentelemetry-collector-contrib), all of this functionality will be optional.
```

> This OTEP is a request for an experimental feature [open-telemetry/opentelemetry-specification#62](https://github.com/open-telemetry/opentelemetry-specification/pull/632). It is intended to seek approval to develop a proof of concept. This means no development will be done inside either the Open Telemetry SDK or the collector. Since this will be implemented in [opentelemetry-go-contrib](https://github.com/open-telemetry/opentelemetry-go-contrib) and [opentelemetry-collector-contrib](https://github.com/open-telemetry/opentelemetry-collector-contrib), all of this functionality will be optional.

Author: If it is acceptable to the OpenTelemetry maintainers, in line with the experiment framework for new features (https://github.com/open-telemetry/opentelemetry-specification/tree/master/experimental), would it be a good idea to have an experimental folder in opentelemetry-proto for experimental protocols? Or is this too invasive?

@SergeyKanzhelev

> This OTEP is a request for an experimental feature [open-telemetry/opentelemetry-specification#62](https://github.com/open-telemetry/opentelemetry-specification/pull/632). It is intended to seek approval to develop a proof of concept. This means no development will be done inside either the Open Telemetry SDK or the collector. Since this will be implemented in [opentelemetry-go-contrib](https://github.com/open-telemetry/opentelemetry-go-contrib) and [opentelemetry-collector-contrib](https://github.com/open-telemetry/opentelemetry-collector-contrib), all of this functionality will be optional.

Reviewer 3: @churanchen I think it is a good idea to also clearly mark the Protobuf messages in opentelemetry-proto as experimental.

> This OTEP is a request for an experimental feature [open-telemetry/opentelemetry-specification#62](https://github.com/open-telemetry/opentelemetry-specification/pull/632). It is intended to seek approval to develop a proof of concept. This means no development will be done inside either the Open Telemetry SDK or the collector. Since this will be implemented in [opentelemetry-go-contrib](https://github.com/open-telemetry/opentelemetry-go-contrib) and [opentelemetry-collector-contrib](https://github.com/open-telemetry/opentelemetry-collector-contrib), all of this functionality will be optional.

Author: I agree. If this is the approach we go through with, we can add an `Experimental_` prefix to all our messages in our past PR.

> This OTEP is a request for an experimental feature [open-telemetry/opentelemetry-specification#62](https://github.com/open-telemetry/opentelemetry-specification/pull/632). It is intended to seek approval to develop a proof of concept. This means no development will be done inside either the Open Telemetry SDK or the collector. Since this will be implemented in [opentelemetry-go-contrib](https://github.com/open-telemetry/opentelemetry-go-contrib) and [opentelemetry-collector-contrib](https://github.com/open-telemetry/opentelemetry-collector-contrib), all of this functionality will be optional.

Author: To clarify, is that what you meant by marking the Protobuf messages?

> This OTEP is a request for an experimental feature [open-telemetry/opentelemetry-specification#62](https://github.com/open-telemetry/opentelemetry-specification/pull/632). It is intended to seek approval to develop a proof of concept. This means no development will be done inside either the Open Telemetry SDK or the collector. Since this will be implemented in [opentelemetry-go-contrib](https://github.com/open-telemetry/opentelemetry-go-contrib) and [opentelemetry-collector-contrib](https://github.com/open-telemetry/opentelemetry-collector-contrib), all of this functionality will be optional.

Reviewer 3: Either that or put them in a separate directory/package (or even do both to be very explicit).

The user, when instrumenting their application, can configure the SDK with the endpoint of their remote configuration service, the associated Resource and a default config we revert to if we fail to read from the configuration service.

> The user must then set up the config service. This MUST be done through the collector, which can be set up to expose an arbitrary configuration service implementation. Depending on implementation, this allows the collector to either act as a stand-alone configuration service, or as a bridge to remote configurations of the user's monitoring and tracing backend by 'translating' the monitoring backend's protocol to comply with the Open Telemetry configuration protocol.

Reviewer 4: +1, @churanchen can you please fix MUST to should or rephrase it in terms - "if user decide to use collector as a config service, than she MUST..."

> The user must then set up the config service. This MUST be done through the collector, which can be set up to expose an arbitrary configuration service implementation. Depending on implementation, this allows the collector to either act as a stand-alone configuration service, or as a bridge to remote configurations of the user's monitoring and tracing backend by 'translating' the monitoring backend's protocol to comply with the Open Telemetry configuration protocol.

Author: Done! We'd originally limited it to the collector to limit scope/possible security problems, but this makes sense to remove.

> The user must then set up the config service. This MUST be done through the collector, which can be set up to expose an arbitrary configuration service implementation. Depending on implementation, this allows the collector to either act as a stand-alone configuration service, or as a bridge to remote configurations of the user's monitoring and tracing backend by 'translating' the monitoring backend's protocol to comply with the Open Telemetry configuration protocol.

Author: Oh shoot, thanks catching this. Should have fixed it everywhere.

## Internal details

In the future, it is intended to add per-metric configuration. For example, this would allow the user to collect 5xx server error counts ever minute, and CPU usage statistics every 10 minutes. The remote configuration protocol was designed with this in mind, meaning that it includes more details than simply the metric collection period.

Our remote configuration protocol will support this call:

```
service DynamicConfig {
  rpc GetConfig (ConfigRequest) returns (ConfigResponse);
}
```

A request to the config service will look like this:

```
message ConfigRequest{

  // Required. The resource for which configuration should be returned.
  opentelemetry.proto.resource.v1.Resource resource = 1;

  // Optional. The value of ConfigResponse.fingerprint for the last configuration
  // a resource received that was successfully applied.
  bytes last_known_fingerprint = 2;
}
```

While the response will look like this:

```
message ConfigResponse {

>   // Optional. The fingerprint associated with this ConfigResponse. Each change

Author: Our intention for the fingerprint field was to have an identifier for configs that the user can set and use if they wish, similar to the metadata field. The idea being that there might be some fields that are nice and useful to include, but that we shouldn't force the end users to have in their configs.

        HR_2 = 7200;
        HR_4 = 14400;
        HR_12 = 43200;
        DAY_1 = 86400;
        DAY_7 = 604800;
      }
      CollectionPeriod period = 3;

      // Optional. Additional opaque metadata associated with the schedule.
      // Interpreting metadata is implementation specific.
      bytes metadata = 4;
    }

    // For this iteration, since we only want one Schedule that applies to all metrics,
    // we will have a restriction that schedules must have a length of 1, and we will
    // not check the patterns when we apply the collection period.
    repeated Schedule schedules = 1;

  }
  MetricConfig metric_config = 2;

  // Dynamic configs specific to trace, like the sampling rate of a resource.
  message TraceConfig {
    // TODO: unimplemented
    // This proposal focuses on the metrics portion, but tracing is a natural
    // candidate for something that might be useful to configure in the future,
    // so this field will be left here for that eventuality.
  }
  TraceConfig trace_config = 3;

  // Optional. The client is suggested to wait this long (in seconds) before
  // pinging the configuration service again.
  int32 suggested_wait_time_sec = 4;
}
```

> The SDK will periodically read a config from the service using GetConfig. If it fails to do so, it will just use either the default config or the most recent successfully read config. If it reads a new config, it will apply it.

Reviewer 4: +1 @churanchen I suggest to add a section in alternative consideration about possible mechanisms of config pull/push. And the decision can be based on perf measurement

> The SDK will periodically read a config from the service using GetConfig. If it fails to do so, it will just use either the default config or the most recent successfully read config. If it reads a new config, it will apply it.

Author: Good points. Mentioned here: https://github.com/open-telemetry/oteps/blob/613b463da66d52bd18a3b2a1b376279ec78e381a/text/0121-config-service.md#prior-art-and-alternatives

Export frequency from the SDK depends on Schedules. There can only be one Schedule for now, which defines the schedule for all metrics. The schedule has a CollectionPeriod, which defines how often metrics are exported.

In the future, we will add per-metric configuration. Each Schedule also has inclusion_patterns and exclusion_patterns. Any metrics that match any of the inclusion_patterns and do not match any of the exclusion_patterns will be exported every CollectionPeriod (ie. every minute). A component will be added that can export metrics that match a pattern from one Schedule at that Schedule's collection period, while exporting other metrics that match the patterns of another Schedule at that other Schedule's collection period.

> The collector will support a new interface for a DynamicConfig service that can be used by an SDK, allowing a custom implementation of the configuration service protocol described above, to act as an optional bridge between an SDK and an arbitrary configuration service. This interface can be implemented as a shim to support accessing remote configurations from arbitrary backends. The collector is configured to expose an endpoint for requests to the DynamicConfig service, and returns results on that endpoint.

Reviewer 5: Great points. Currently in our prototype, the service is implemented as an extension on the collector (like zpages, healthcheck). If enabled, the user has the option of specifying one of two (or more, potentially) backends from which the collector gets its configs:
*  file backend: a local file containing the metric and trace configs is served. This option corresponds to the usage you mentioned in which the collector itself is the configuration service. The file is actively monitored, so any updates to the file will be propagated to clients upon their next update request.
* remote backend: the collector receives config updates from a remote backend that implements the server-side of the gRPC service. This remote backend could be another collector, or it can be third-party. In this way, the remote backend fufills the bridging concept.

I like your idea of implementing the process as a pipeline. We chose an extension originally as a simple way to cleanly develop a prototype, but as the implementation grows, it may be nice to organize it into a full pipeline.

For specific details, feel free to check out a [PR introducing the extension](https://github.com/vmingchen/opentelemetry-collector-contrib/pull/1) and [another that adds the remote backend](https://github.com/vmingchen/opentelemetry-collector-contrib/pull/2). Both are made on our intern host's fork of the project, and will be moved to the upstream contrib repo when finished.

Does this approach sound viable? Let me know what you think!

## Trade-offs and mitigations

This feature will be implemented purely as an experiment, to demonstrate its viability and usefulness. More investigation can be done after a rough prototype is demonstrated.

As mentioned [here](https://github.com/open-telemetry/opentelemetry-proto/pull/155#issuecomment-640582048), the configuration service can be a potential attack vector for an application instrumented with Open Telemetry, depending on what we allow in the protocol. This can be mitigated since this proposal only allows the SDK to read from a configuration service managed by the collector. Additionally, we can highlight in the remote configuration protocol that for future changes, caution is needed in terms of the sorts of configurations we allow.

> ## Prior art and alternatives

Reviewer 4: +1 Good thing to explore in this experiment

> ## Prior art and alternatives

Author: We did think about a push mechanism. Our rationale for a polling mechanism is that it might be more scalable if you have a central configuration service that is read from many processes, rather than the service having to push to many processes. However, this is something that definitely needs a harder examination.

> One way to configure metric export schedules is to, instead of pushing the metrics, use a pull mechanism to have metrics pulled whenever wanted. This has been implemented for [Prometheus](https://github.com/open-telemetry/opentelemetry-go/pull/751).

Reviewer 2: This is an alternative to the current push-based approach to metrics, but not really an alternative to the proposal here. Echoing @tigrannajaryan's comment above, it sounds like the important alternative to note here is pushing vs. polling the config.

On its face, It's pretty surprising that we'd ask end users to implement a remote configuration service which the collector (or SDK) would poll, instead of adding something like a `setConfig` API method to the collector. Why did you decide on polling here?

Author: Our primary rationale was the large-scale use case where the collector would be responsible for pushing configs to many instrumented processes. It seemed more scalable to have the SDK read from the configuration service.

In a situation where the collector is extended to act as a configuration service or act as a proxy for remote configuration built into a monitoring backend, hopefully the end user won't be need to do much extra work relative to setting up a mechanism that calls a `setConfig` method on the collector.

The trade-offs for pushing vs. polling is definitely something worth thinking about though.

The alternative is to stick with the status quo, where the SDK has a [fixed collection period](https://github.com/open-telemetry/opentelemetry-go/blob/34bd99896311a81cf843475779cae2e1c05e6257/sdk/metric/controller/push/push.go#L72-L76) and a fixed sampling rate.

## Open questions

- As mentioned [here](https://github.com/open-telemetry/opentelemetry-proto/pull/155#issuecomment-640582048). what happens if a malicious/accidental config change overwhelms the application/monitoring system? Is it the responsibility of the user to be cautious while making config changes? Should we automatically decrease telemetry exporting if we can detect performance problems?

> ## Future possibilities

Reviewer 4: I like your suggestion on thinking of these pipelines separately. If we do so - it is quite logical, that collector may be configured to participate in config pipeline. But the default mode may be a simple proxy of configs back to whoever is asking for it.

> ## Future possibilities

Author: As far as our prototype works so far, we intend the collector to only be a proxy.

For the future, if the collector can be configured, maybe individual collectors should also be labelled (if they are not already), in the way resources are currently. There may be a large-scale use case where there are "layers" of collectors, and it's necessary to configure individual collectors. Then, a label given to each configured process can also be used to distinguish between SDK and collector.

          string starts_with = 2;  // prefix-matches the metric name
        }
      }

      // Metrics with names that match at least one rule in the inclusion_patterns are
      // targeted by this schedule. Metrics that match at least one rule from the
      // exclusion_patterns are not targeted for this schedule, even if they match an
      // inclusion pattern.

      // For this iteration, since we only want one Schedule that applies to all metrics,
      // we will not check the inclusion_patterns and exclusion_patterns.
      repeated Pattern inclusion_patterns = 1;
      repeated Pattern exclusion_patterns = 2;

      // CollectionPeriod describes the sampling period for each metric. All
      // larger units are divisible by all smaller ones.
      enum CollectionPeriod {
        NONE = 0;  // For non-periodic data (client sends points whenever)
        SEC_1 = 1;
        SEC_5 = 5;
        SEC_10 = 10;
        SEC_30 = 30;
        MIN_1 = 60;
        MIN_5 = 300;
        MIN_10 = 600;
        MIN_30 = 1800;
        HR_1 = 3600;
        HR_2 = 7200;
        HR_4 = 14400;
        HR_12 = 43200;
        DAY_1 = 86400;
        DAY_7 = 604800;
      }

>       CollectionPeriod period = 3;

Reviewer 6: Is this enum needed, if the number is in seconds? Given that the values are in seconds, it implies that unlisted values (e.g., 2 days == 172800) may be used. If unlisted values can be used, do we need these constants?

How is NONE meant to be used?

Author: The enum is used to enforce the property that all larger periods must be multiples of the smaller ones. This might be useful in that we can batch the export of metrics with less frequent CollectionPeriods with those with more frequent CollectionPeriods. 

Just using an int definitely definitely provides more flexibility. Let me know what you think!

So I deleted the comment left with `NONE`. The intention of `NONE` is just to say "do not export". Might be useful for the user to have an option they can toggle to stop exports for the user. `NONE` can be gotten rid of if it seems unnecessary.

Reviewer 7: Since the server supplies this configuration to the SDK, it can have whichever validation on multiples that it wants, without having that reflected in the protocol schema. So +1 to just have a duration field. The field name should reflect that it is in seconds.

Author: Done!

      // Optional. Additional opaque metadata associated with the schedule.
      // Interpreting metadata is implementation specific.

>       bytes metadata = 4;

Reviewer 6: This looks difficult to use. How do you see it being used?

Author: This is a sort of catch-all field for metadata that is used by a specific monitoring back-end implementation but that we do not specifically include in the protocol. Say a monitoring system has the notion of network traffic priority. It might be useful to relay this to the instrumented application, or for the instrumented application to relay this back to the back-end. It's up to the user if/how they want to use this field.

Reviewer 7: I would prefer to remove this. Some of the scenarios you describe are general enough and could be expressed via proper schema in the future. For pure custom vendor extensions I don't think `bytes metadata` is a good format. Doesn't proto allow custom Any types? And does the custom extension even need to be part of this proto? It will require custom SDKs, and there are other ways custom SDKs can access this.

Author: Removed! This is still something we might be interested in later, but it's not necessary for a proof-of-concept.

# A Dynamic Configuration Service for the SDK

This proposal is a request to develop a prototype to configure metric collection periods. Per-metric and tracing configuration is also intended to be added, with details left for a later iteration.

It is related to [this pull request](https://github.com/open-telemetry/opentelemetry-proto/pull/155)

## Motivation

During normal use, users may wish to collect metrics every 10 minutes. Later, while investigating a production issue, the same user could easily increase information available for debugging by reconfiguring some of their processes to collect metrics every 30 seconds. Because this change is centralized and does not require redeploying with new configurations, there is lower friction and risk in updating the configurations.

## Explanation

> This OTEP is a request for an experimental feature [open-telemetry/opentelemetry-specification#62](https://github.com/open-telemetry/opentelemetry-specification/pull/632). It is intended to seek approval to develop a proof of concept. This means no development will be done inside either the OpenTelemetry SDK or the collector. Since this will be implemented in [opentelemetry-go-contrib](https://github.com/open-telemetry/opentelemetry-go-contrib) and [opentelemetry-collector-contrib](https://github.com/open-telemetry/opentelemetry-collector-contrib), all of this functionality will be optional.

Reviewer 7: ```suggestion
This OTEP is a proposal for an experimental feature [open-telemetry/opentelemetry-specification#62](https://github.com/open-telemetry/opentelemetry-specification/pull/632), to be developed as a proof of concept. This means no development will be done inside either the OpenTelemetry SDK or the collector. Since this will be implemented in [opentelemetry-go-contrib](https://github.com/open-telemetry/opentelemetry-go-contrib) and [opentelemetry-collector-contrib](https://github.com/open-telemetry/opentelemetry-collector-contrib), all of this functionality will be optional.
```

> The user, when instrumenting their application, can configure the SDK with the endpoint of their remote configuration service, the associated Resource and a default config we revert to if we fail to read from the configuration service.

Reviewer 7: ```suggestion
The user, when instrumenting their application, can configure the SDK with the endpoint of their remote configuration service, the associated Resource, and a default config to be used if fail to read from the configuration service.
```

The user must then set up the config service. This can be done through the collector, which can be set up to expose an arbitrary configuration service implementation. Depending on implementation, this allows the collector to either act as a stand-alone configuration service, or as a bridge to remote configurations of the user's monitoring and tracing backend by 'translating' the monitoring backend's protocol to comply with the OpenTelemetry configuration protocol.

## Internal details

> In the future, it is intended to add per-metric configuration. For example, this would allow the user to collect 5xx server error counts ever minute, and CPU usage statistics every 10 minutes. The remote configuration protocol was designed with this in mind, meaning that it includes more details than simply the metric collection period.

Reviewer 7: ```suggestion
In the future, we intend to add per-metric configuration. For example, this would allow the user to collect 5xx server error counts ever minute, and CPU usage statistics every 10 minutes. The remote configuration protocol was designed with this in mind, meaning that it includes more details than simply the metric collection period.
```

Our remote configuration protocol will support this call:

```
service DynamicConfig {

>   rpc GetConfig (ConfigRequest) returns (ConfigResponse);

Reviewer 7: I don't particularly like bundling all configs like that into a single RPC call. For example, Uber currently instruments all apps with OpenTracing and Tally for traces/metrics respectively. We may have a migration path to swap tracing instrumentation for OpenTelemetry, but replacing Tally instrumentation is going to be a huge effort that's not worth it. So we would only need remote configurability for the tracing part of OTEL SDK.

I would prefer separating these into different services, DynamicTracingConfig, DynamicMetricsConfig, etc.

Author: I definitely agree that users should not be forced into configuring both metrics and tracing if they only need to configure one. What if the protocol supported having MetricConfig and TraceConfig be optional? Uber, for example, could choose to remotely configure tracing, but not have a metrics config at all.

I worry in the future, if remote configuration is adopted, that it may get unwieldy if we have a DynamicTracingConfig, DynamicMetricsConfig, DynamicLoggingConfig (if OpenTelemetry supports this in the future), etc. where if a user wants configure multiple parts of the SDK, multiple calls need to be made to a configuration service.

What do you think? Does this sound okay, or would you still prefer having separate services?

Reviewer 7: We need to see what other people think. Personally, I am not too bothered by multiple calls. It also allows different TTLs for different types of configuration.

Author: Hmm, how about we change DynamicConfig -> DynamicMetricsConfig, and just leave out DynamicTracingConfig for this OTEP. Our OTEP is more about developing a proof-of-concept for the idea of remote configuration, and it doesn't go into detail about tracing. If remote configuration gets adopted into OpenTelemetry, a DynamicTracingConfig service can be added later.

Reviewer 7: +1

Author: Done!

}
```

A request to the config service will look like this:

```
message ConfigRequest{

  // Required. The resource for which configuration should be returned.
  opentelemetry.proto.resource.v1.Resource resource = 1;

  // Optional. The value of ConfigResponse.fingerprint for the last configuration

>   // a resource received that was successfully applied.

Reviewer 7: ```suggestion
  // that the caller received and successfully applied.
```

  message MetricConfig {

    // A Schedule is used to apply a particular scheduling configuration to
    // a metric. If a metric name matches a schedule's patterns, then the metric
    // adopts the configuration specified by the schedule.

    message Schedule {

      // A light-weight pattern that can match 1 or more
      // metrics, for which this schedule will apply. The string is used to
      // match against metric names. It should not exceed 100k characters.
      message Pattern {
        oneof match {
          string equals = 1;       // matches the metric name exactly
          string starts_with = 2;  // prefix-matches the metric name
        }
      }

      // Metrics with names that match at least one rule in the inclusion_patterns are
      // targeted by this schedule. Metrics that match at least one rule from the
      // exclusion_patterns are not targeted for this schedule, even if they match an
      // inclusion pattern.

      // For this iteration, since we only want one Schedule that applies to all metrics,
      // we will not check the inclusion_patterns and exclusion_patterns.
      repeated Pattern inclusion_patterns = 1;

>       repeated Pattern exclusion_patterns = 2;

Reviewer 7: nits: 
  1. because exclusion patterns have higher priority per the comment above, I would move it to the first position.
  2. I am not sure this grammar is expressive enough. Is there any prior art for this?

Author: 1. Changed!
2. The choice to use exact match and starts with was made because they are more efficient to apply and clearer to read than more expressive alternatives, and have in practice been the only required pattern types for specifying metric names in Google's internal metric collection configuration protocol. However we are open to also adding a more expressive grammar such as regex or glob as a third option for a pattern type, so that monitoring backends can optionally choose to support a more expressive grammar without affecting efficiency for the simpler pattern types.
